Assets/Scripts/Hit.cs

```diff
@@ -0,0 +1,177 @@
+using UnityEngine;
+using UnityEngine.Events;
+using System.Collections.Generic;
+using System.Linq;
+
+// Interface for objects that can hit others
+public interface IHitter
+{
+    void ApplyDamage(GameObject target);
+}
+
+public class Hit : MonoBehaviour, IHitter
+{
+    [Header("Damage Settings")]
+    [SerializeField] private float damage = 10f;
+
+    [Header("Hit Behavior")]
+    [SerializeField] private bool destroyOnHit = false;
+    [SerializeField] private int hitMaxTimes = 1;
+    [SerializeField] private float hitCooldown = 0.5f;
+    [SerializeField] private bool debugMode = false;
+
+    [Header("Events")]
+    public UnityEvent<GameObject> OnHitObject;
+    public UnityEvent<GameObject> OnBeforeDestroy;
+
+    private Dictionary<GameObject, int> hitCounts = new Dictionary<GameObject, int>(); // Track hit counts
+    private Dictionary<GameObject, float> hitCooldowns = new Dictionary<GameObject, float>();
+
+    public void Configure(float damage, bool destroyOnHit, int hitMaxTimes, float hitCooldown)
+    {
+        this.damage = damage;
+        this.destroyOnHit = destroyOnHit;
+        this.hitMaxTimes = hitMaxTimes;
+        this.hitCooldown = hitCooldown;
+
+        if (debugMode)
+            Debug.Log($"[Hit:{gameObject.name}] Configured with damage={damage}, destroyOnHit={destroyOnHit}, hitMaxTimes={hitMaxTimes}, cooldown={hitCooldown}");
+    }
+
+    public void SetDamage(float damage)
+    {
+        this.damage = damage;
+    }
+
+    public void SetDestroyOnHit(bool destroyOnHit)
+    {
+        this.destroyOnHit = destroyOnHit;
+    }
+
+    public void SetHitMaxTimes(int hitMaxTimes)
+    {
+        this.hitMaxTimes = hitMaxTimes;
+    }
+
+    public void SetHitCooldown(float hitCooldown)
+    {
+        this.hitCooldown = hitCooldown;
+    }
+
+    public void SetDebugMode(bool debugMode)
+    {
+        this.debugMode = debugMode;
+    }
+
+    private void Start()
+    {
+        if (debugMode)
+            Debug.Log($"[Hit:{gameObject.name}] Initialized with damage={damage}, destroyOnHit={destroyOnHit}, hitMaxTimes={hitMaxTimes}");
+    }
+
+    void Update()
+    {
+        // Manage cooldowns
+        List<GameObject> objectsToRemove = new List<GameObject>();
```

Comment:
i think rather than doing this on each frame we can use pooling to clean this up - as it could potentially get quite large. Same as above